File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)

        while(left <= right):

            mid = (left + right) // 2

            hourse_needed = 0

            for i in piles:
                hourse_needed = hourse_needed + ( (i + (mid - 1)) // mid) 
            
            if(hourse_needed <= h):
                right = mid - 1
            else:
                left = mid + 1

        return left

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        
    	int left = 1;
    	
    	int max = piles[0];
    	for(int pile : piles) {
    		if(pile > max) {
    			max = pile;
    		}
    	}
    	int right = max;
    	
    	while(left <= right) {
    		int mid = (left + right) / 2;
    		
    		long eating_time = 0;
    		
    		for(int i=0; i < piles.length; i++) {
    			eating_time = eating_time + (piles[i] + mid - 1) / mid;
    		}
    		
    		if(eating_time <= h) {
    			right = mid - 1;
    		}
    		else {
    			left = mid + 1;
    		}
    	}
    	return left;
    }
}